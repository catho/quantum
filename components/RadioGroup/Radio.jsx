import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Label, HiddenInput, shadow } from '../shared';
import { colors, spacing } from '../shared/theme';

const RadioMark = styled.span`
  border-radius: 50%;
  box-sizing: border-box;
  display: inline-block;
  height: 16px;
  position: relative;
  top: 4px;
  width: 16px;

  ${({
    theme: {
      colors: {
        neutral: { 0: white, 500: gray },
      },
      spacing: { xxsmall },
    },
  }) => `
    background-color: ${white};
    border: 1.5px solid ${gray};
    margin-right: ${xxsmall}px;
  `}

  :after {
    border-radius: 50%;
    content: '';
    display: none;
    height: 50%;
    left: 25%;
    position: absolute;
    top: 25%;
    width: 50%;

    background-color: ${({
      theme: {
        colors: {
          primary: { 500: primary },
        },
      },
    }) => primary};
  }
`;

const RadioLabel = styled(Label)`
  cursor: pointer;
  user-select: none;

  ${HiddenInput} {
    :checked {
      ~ ${RadioMark} {
        ${({
          theme: {
            colors: {
              primary: { 500: primary },
            },
          },
        }) => `
          border-color: ${primary};
          :after {
            background-color: ${primary};
            display: block;
          }
        `}
      }
    }

    :focus {
      ~ ${RadioMark} {
        ${({ theme }) => {
          const {
            colors: {
              primary: { 500: primary },
            },
          } = theme;

          return `
            border-color: ${primary};
            ${shadow(5, primary)({ theme })}
          `;
        }}
      }
    }
  }

  :hover,
  :focus {
    ${RadioMark} {
      ${({ theme }) => {
        const {
          colors: {
            primary: { 500: primary },
          },
        } = theme;

        return `
          border-color: ${primary};
          ${shadow(5, primary)({ theme })}
        `;
      }}
    }
  }

  ${({ theme, error }) => {
    const {
      colors: {
        error: { 500: errorColor },
      },
    } = theme;

    return (
      error &&
      `
      ${HiddenInput} {
        :checked {
          ~ ${RadioMark} {
            border-color: ${errorColor};
            :after {
              background-color: ${errorColor};
              display: block;
            }
          }
        }
        :focus {
          ~ ${RadioMark} {
            border-color: ${errorColor};
            ${shadow(5, errorColor)({ theme })}
          }
        }
      }

      ${RadioMark} {
        border-color: ${errorColor};
      }

      :hover,
      :focus {
        ${RadioMark} {
          border-color: ${errorColor};
          ${shadow(5, errorColor)({ theme })}
        }
      }
    `
    );
  }}

  ${({
    disabled,
    theme: {
      colors: {
        neutral: { 500: neutral500, 300: neutral300 },
        error: { 500: errorColor },
      },
    },
  }) =>
    disabled &&
    `
    color: ${neutral500};

    :hover {
      cursor: not-allowed;
    }

    ${HiddenInput} {
      :disabled {
        ~ ${RadioMark} {
          background-color: ${neutral300};
        }
      }

      :checked {
        ~ ${RadioMark} {
          border-color: ${errorColor};
          :after {
            background-color: ${errorColor};
            display: block;
          }
        }
      }

      :checked:disabled {
        background-color: ${neutral500};
        border-color: ${neutral500};
        ~ ${RadioMark}:after {
          background-color: ${neutral500};
        }
      }

      :focus {
        ~ ${RadioMark} {
          border-color: ${neutral500};
          box-shadow: none;
        }
      }
    }

    :hover, :focus {
      ${RadioMark} {
        border-color: ${neutral500};
        box-shadow: none;
      }
    }
  `}

  ${HiddenInput}:checked:disabled ~ ${RadioMark} {
    ${({
      theme: {
        colors: {
          neutral: { 0: white, 500: black },
        },
      },
    }) => `
      background-color: ${black};
      border-color: ${black};
      box-shadow: inset 0 0 0 3.5px ${white};
    `}
  }
`;

const Radio = ({
  children,
  label,
  error,
  disabled,
  onChange,
  value,
  theme,
  ...rest
}) => (
  <RadioLabel error={error} disabled={disabled} theme={theme}>
    <HiddenInput
      type="radio"
      disabled={disabled}
      onChange={e => onChange({ value, label }, e)}
      value={value}
      {...rest}
    />
    <RadioMark theme={theme} />
    {children || label}
  </RadioLabel>
);

Radio.displayName = 'RadioGroup.Radio';

Radio.propTypes = {
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  children: PropTypes.string,
  label: PropTypes.string,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  theme: PropTypes.shape({
    colors: PropTypes.object,
    spacing: PropTypes.object,
  }),
};

Radio.defaultProps = {
  disabled: false,
  error: false,
  children: undefined,
  label: undefined,
  onChange: () => {},
  theme: { colors, spacing },
};

export default Radio;
