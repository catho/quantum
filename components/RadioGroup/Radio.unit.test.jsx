import React from 'react';
import renderer from 'react-test-renderer';
import Radio from './Radio';

describe('<RadioGroup.Radio />', () => {
  describe('snapshot', () => {
    it('simple', () => {
      const component = <Radio value="Foo" label="Foo" />;
      expect(renderer.create(component).toJSON()).toMatchSnapshot();
    });

    it('disabled', () => {
      const component = <Radio value="Foo" label="Foo" disabled />;
      expect(renderer.create(component).toJSON()).toMatchSnapshot();
    });

    it('error', () => {
      const component = <Radio value="Foo" label="Foo" error />;
      expect(renderer.create(component).toJSON()).toMatchSnapshot();
    });

    it('children', () => {
      const component = <Radio value="Foo">Foo</Radio>;
      expect(renderer.create(component).toJSON()).toMatchSnapshot();
    });

    it('children with element', () => {
      const component = (
        <Radio value="Foo">
          <span>Foo</span>
        </Radio>
      );
      expect(renderer.create(component).toJSON()).toMatchSnapshot();
    });

    it('required', () => {
      const component = (
        <Radio value="Foo" required>
          Foo
        </Radio>
      );
      expect(renderer.create(component).toJSON()).toMatchSnapshot();
    });
  });
});
