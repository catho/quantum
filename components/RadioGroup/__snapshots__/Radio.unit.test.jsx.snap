// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup.Radio /> snapshot children 1`] = `
.c0 {
  display: block;
  margin-bottom: 5px;
}

.c3 {
  border: 0;
  height: 0;
  margin: 0;
  opacity: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 0;
}

.c5 {
  border-radius: 50%;
  box-sizing: border-box;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  height: 24px;
  position: relative;
  top: 6px;
  width: 24px;
  background-color: #ffffff;
  border: 2px solid #999999;
  margin-right: 8px;
}

.c5:after {
  border-radius: 50%;
  content: '';
  display: none;
  height: 60%;
  left: 20%;
  position: absolute;
  top: 20%;
  width: 60%;
  background-color: #1250C4;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 .c2:checked ~ .c4 {
  border-color: #1250C4;
}

.c1 .c2:checked ~ .c4:after {
  background-color: #1250C4;
  display: block;
}

.c1 .c2:focus ~ .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1:hover .c4,
.c1:focus .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1 .c2:checked:disabled ~ .c4 {
  background-color: #999999;
  border-color: #999999;
  box-shadow: inset 0 0 0 3.5px #ffffff;
}

<label
  className="c0 c1"
  disabled={false}
>
  <input
    className="c2 c3"
    disabled={false}
    onChange={[Function]}
    required={false}
    type="radio"
    value="Foo"
  />
  <span
    className="c4 c5"
  />
  <span>
    Foo
  </span>
</label>
`;

exports[`<RadioGroup.Radio /> snapshot children with element 1`] = `
.c0 {
  display: block;
  margin-bottom: 5px;
}

.c3 {
  border: 0;
  height: 0;
  margin: 0;
  opacity: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 0;
}

.c5 {
  border-radius: 50%;
  box-sizing: border-box;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  height: 24px;
  position: relative;
  top: 6px;
  width: 24px;
  background-color: #ffffff;
  border: 2px solid #999999;
  margin-right: 8px;
}

.c5:after {
  border-radius: 50%;
  content: '';
  display: none;
  height: 60%;
  left: 20%;
  position: absolute;
  top: 20%;
  width: 60%;
  background-color: #1250C4;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 .c2:checked ~ .c4 {
  border-color: #1250C4;
}

.c1 .c2:checked ~ .c4:after {
  background-color: #1250C4;
  display: block;
}

.c1 .c2:focus ~ .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1:hover .c4,
.c1:focus .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1 .c2:checked:disabled ~ .c4 {
  background-color: #999999;
  border-color: #999999;
  box-shadow: inset 0 0 0 3.5px #ffffff;
}

<label
  className="c0 c1"
  disabled={false}
>
  <input
    className="c2 c3"
    disabled={false}
    onChange={[Function]}
    required={false}
    type="radio"
    value="Foo"
  />
  <span
    className="c4 c5"
  />
  <span>
    <span>
      Foo
    </span>
  </span>
</label>
`;

exports[`<RadioGroup.Radio /> snapshot disabled 1`] = `
.c0 {
  display: block;
  margin-bottom: 5px;
}

.c3 {
  border: 0;
  height: 0;
  margin: 0;
  opacity: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 0;
}

.c5 {
  border-radius: 50%;
  box-sizing: border-box;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  height: 24px;
  position: relative;
  top: 6px;
  width: 24px;
  background-color: #ffffff;
  border: 2px solid #999999;
  margin-right: 8px;
}

.c5:after {
  border-radius: 50%;
  content: '';
  display: none;
  height: 60%;
  left: 20%;
  position: absolute;
  top: 20%;
  width: 60%;
  background-color: #1250C4;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #999999;
}

.c1 .c2:checked ~ .c4 {
  border-color: #1250C4;
}

.c1 .c2:checked ~ .c4:after {
  background-color: #1250C4;
  display: block;
}

.c1 .c2:focus ~ .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1:hover .c4,
.c1:focus .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1:hover {
  cursor: not-allowed;
}

.c1 .c2:disabled ~ .c4 {
  background-color: #e0e0e0;
}

.c1 .c2:checked ~ .c4 {
  border-color: #ec584e;
}

.c1 .c2:checked ~ .c4:after {
  background-color: #ec584e;
  display: block;
}

.c1 .c2:checked:disabled {
  background-color: #999999;
  border-color: #999999;
}

.c1 .c2:checked:disabled ~ .c4:after {
  background-color: #999999;
}

.c1 .c2:focus ~ .c4 {
  border-color: #999999;
  box-shadow: none;
}

.c1:hover .c4,
.c1:focus .c4 {
  border-color: #999999;
  box-shadow: none;
}

.c1 .c2:checked:disabled ~ .c4 {
  background-color: #999999;
  border-color: #999999;
  box-shadow: inset 0 0 0 3.5px #ffffff;
}

<label
  className="c0 c1"
  disabled={true}
>
  <input
    className="c2 c3"
    disabled={true}
    onChange={[Function]}
    required={false}
    type="radio"
    value="Foo"
  />
  <span
    className="c4 c5"
  />
  <span>
    Foo
  </span>
</label>
`;

exports[`<RadioGroup.Radio /> snapshot error 1`] = `
.c0 {
  display: block;
  margin-bottom: 5px;
}

.c3 {
  border: 0;
  height: 0;
  margin: 0;
  opacity: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 0;
}

.c5 {
  border-radius: 50%;
  box-sizing: border-box;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  height: 24px;
  position: relative;
  top: 6px;
  width: 24px;
  background-color: #ffffff;
  border: 2px solid #999999;
  margin-right: 8px;
}

.c5:after {
  border-radius: 50%;
  content: '';
  display: none;
  height: 60%;
  left: 20%;
  position: absolute;
  top: 20%;
  width: 60%;
  background-color: #1250C4;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 .c2:checked ~ .c4 {
  border-color: #1250C4;
}

.c1 .c2:checked ~ .c4:after {
  background-color: #1250C4;
  display: block;
}

.c1 .c2:focus ~ .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1:hover .c4,
.c1:focus .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1 .c2:checked ~ .c4 {
  border-color: #e91b0c;
}

.c1 .c2:checked ~ .c4:after {
  background-color: #e91b0c;
  display: block;
}

.c1 .c2:focus ~ .c4 {
  border-color: #e91b0c;
  box-shadow: 0px 3px 5px -1px rgba(233,27,12,0.2),0px 5px 8px 0px rgba(233,27,12,0.14),0px 1px 14px 0px rgba(233,27,12,0.12);
}

.c1 .c4 {
  border-color: #e91b0c;
}

.c1:hover .c4,
.c1:focus .c4 {
  border-color: #e91b0c;
  box-shadow: 0px 3px 5px -1px rgba(233,27,12,0.2),0px 5px 8px 0px rgba(233,27,12,0.14),0px 1px 14px 0px rgba(233,27,12,0.12);
}

.c1 .c2:checked:disabled ~ .c4 {
  background-color: #999999;
  border-color: #999999;
  box-shadow: inset 0 0 0 3.5px #ffffff;
}

<label
  className="c0 c1"
  disabled={false}
>
  <input
    className="c2 c3"
    disabled={false}
    onChange={[Function]}
    required={false}
    type="radio"
    value="Foo"
  />
  <span
    className="c4 c5"
  />
  <span>
    Foo
  </span>
</label>
`;

exports[`<RadioGroup.Radio /> snapshot required 1`] = `
.c0 {
  display: block;
  margin-bottom: 5px;
}

.c3 {
  border: 0;
  height: 0;
  margin: 0;
  opacity: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 0;
}

.c5 {
  border-radius: 50%;
  box-sizing: border-box;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  height: 24px;
  position: relative;
  top: 6px;
  width: 24px;
  background-color: #ffffff;
  border: 2px solid #999999;
  margin-right: 8px;
}

.c5:after {
  border-radius: 50%;
  content: '';
  display: none;
  height: 60%;
  left: 20%;
  position: absolute;
  top: 20%;
  width: 60%;
  background-color: #1250C4;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 .c2:checked ~ .c4 {
  border-color: #1250C4;
}

.c1 .c2:checked ~ .c4:after {
  background-color: #1250C4;
  display: block;
}

.c1 .c2:focus ~ .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1:hover .c4,
.c1:focus .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1 .c2:checked:disabled ~ .c4 {
  background-color: #999999;
  border-color: #999999;
  box-shadow: inset 0 0 0 3.5px #ffffff;
}

<label
  className="c0 c1"
  disabled={false}
>
  <input
    className="c2 c3"
    disabled={false}
    onChange={[Function]}
    required={true}
    type="radio"
    value="Foo"
  />
  <span
    className="c4 c5"
  />
  <span>
    Foo
  </span>
</label>
`;

exports[`<RadioGroup.Radio /> snapshot simple 1`] = `
.c0 {
  display: block;
  margin-bottom: 5px;
}

.c3 {
  border: 0;
  height: 0;
  margin: 0;
  opacity: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 0;
}

.c5 {
  border-radius: 50%;
  box-sizing: border-box;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  height: 24px;
  position: relative;
  top: 6px;
  width: 24px;
  background-color: #ffffff;
  border: 2px solid #999999;
  margin-right: 8px;
}

.c5:after {
  border-radius: 50%;
  content: '';
  display: none;
  height: 60%;
  left: 20%;
  position: absolute;
  top: 20%;
  width: 60%;
  background-color: #1250C4;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 .c2:checked ~ .c4 {
  border-color: #1250C4;
}

.c1 .c2:checked ~ .c4:after {
  background-color: #1250C4;
  display: block;
}

.c1 .c2:focus ~ .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1:hover .c4,
.c1:focus .c4 {
  border-color: #1250C4;
  box-shadow: 0px 3px 5px -1px rgba(18,80,196,0.2),0px 5px 8px 0px rgba(18,80,196,0.14),0px 1px 14px 0px rgba(18,80,196,0.12);
}

.c1 .c2:checked:disabled ~ .c4 {
  background-color: #999999;
  border-color: #999999;
  box-shadow: inset 0 0 0 3.5px #ffffff;
}

<label
  className="c0 c1"
  disabled={false}
>
  <input
    className="c2 c3"
    disabled={false}
    onChange={[Function]}
    required={false}
    type="radio"
    value="Foo"
  />
  <span
    className="c4 c5"
  />
  <span>
    Foo
  </span>
</label>
`;
